libflux/src/semantic/sub.rs

```diff
@@ -0,0 +1,55 @@
+use crate::semantic::types::{MonoType, Tvar};
+use std::{collections::HashMap, fmt};
+
+// A substitution is a function that takes a monotype as input
+// and returns an equivalent monotype as output.
+//
+// More formally a substitution is a function S: X -> Y such that
+// X = {set of all monotypes} and Y is a subset of X.
+//
+// A substitution is therefore a function that can be applied
+// to monotypes.
+//
+pub trait Substitutable {
+    fn apply(self, sub: &Subst) -> Self;
+}
+
+// A Substitution is implemented as a map from type variables
+// to monotypes.
+//
+// Note this implementation implies that substitutions are
+// idempotent. Given a monotype x, and a substitution s, s(s(x)) = s(x).
+//
+#[derive(Debug)]
+pub struct Subst(HashMap<Tvar, MonoType>);
```

Comment:
Same as above. I implemented the `fmt::Display` trait for substitutions. If it weren't for that implementation you are right, this could be just an alias. The Go code has methods for displaying substitutions and constraint sets. I found them to be quite useful when debugging.